### PR TITLE
Change marathon app stop into Go

### PR DESCRIPTION
**Pull request by br-lewis**
Addresses: https://jira.mesosphere.com/browse/DCOS-60277

### CODE REVIEW

**Comment by br-lewis** [comment]
@mesosphere-ci run mac integration tests

**Comment by br-lewis** [comment]
Looks like mac test failed due to transient issue not related to this PR

`STDERR: URL [https://34.205.33.156/service/marathon/v2/deployments] is unreachable.`

File: pkg/cmd/marathon/app/app_stop_test.go [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+		}
+	}))
+
+	out := new(bytes.Buffer)
+	env := mock.NewEnvironment()
+	env.Out = out
+	ctx := mock.NewContext(env)
+	cluster := config.NewCluster(nil)
+	cluster.SetURL(ts.URL)
+	ctx.SetCluster(cluster)
```
Comment (janisz) [inline pkg/cmd/marathon/app/app_stop_test.go]: extract to function

File: pkg/cmd/marathon/app/app_stop_test.go [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+	"github.com/dcos/dcos-cli/pkg/mock"
+	goMarathon "github.com/gambol99/go-marathon"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func TestAppStop(t *testing.T) {
+	app := struct {
+		App goMarathon.Application `json:"app"`
+	}{
```
Comment (janisz) [inline pkg/cmd/marathon/app/app_stop_test.go]: Create a type and use in all tests
```
type app struct {
	App goMarathon.Application `json:"app"`
}
```

File: pkg/cmd/marathon/app/app_stop_test.go [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+	env := mock.NewEnvironment()
+	env.Out = out
+	ctx := mock.NewContext(env)
+	cluster := config.NewCluster(nil)
+	cluster.SetURL(ts.URL)
+	ctx.SetCluster(cluster)
+
+	err := appStop(ctx, "/test", false)
+	assert.EqualError(t, err, "changes blocked: deployment already in progress for app")
+}
```
Comment (janisz) [inline pkg/cmd/marathon/app/app_stop_test.go:134]: How about this. It's easier to read and easy to turn into table test

```go
package app

import (
	"bytes"
	"encoding/json"
	"net/http"
	"net/http/httptest"
	"testing"

	"github.com/dcos/dcos-cli/pkg/config"
	"github.com/dcos/dcos-cli/pkg/mock"
	goMarathon "github.com/gambol99/go-marathon"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

type app struct {
	App goMarathon.Application `json:"app"`
}

const testID = "test"

func TestAppStop(t *testing.T) {
	app := app{
		App: goMarathon.Application{
			ID:        testID,
			Instances: intPointer(3),
		},
	}
	deployment := goMarathon.DeploymentID{
		DeploymentID: "5ed4c0c5-9ff8-4a6f-a0cd-f57f59a34b43",
		Version:      "2015-09-29T15:59:51.164Z",
	}

	ts := router(
		func(w http.ResponseWriter, r *http.Request) { encode(w, app) },
		func(w http.ResponseWriter, r *http.Request) { encode(w, deployment) },
	)

	ctx, out := newContext(ts)

	err := appStop(ctx, "/test", false)
	require.NoError(t, err)

	assert.Equal(t, "Created deployment 5ed4c0c5-9ff8-4a6f-a0cd-f57f59a34b43\n", out.String())

}

func TestAppStopWhenAppIsStopped(t *testing.T) {
	app := app{
		App: goMarathon.Application{
			ID:        testID,
			Instances: intPointer(0),
		},
	}

	ts := router(
		func(w http.ResponseWriter, r *http.Request) { encode(w, app) },
		nil,
	)

	ctx, _ := newContext(ts)

	err := appStop(ctx, "/test", false)
	assert.EqualError(t, err, "app '/test' already stopped: 0 instances")
}

func TestAppStopWhenAppDoesntExist(t *testing.T) {
	ts := router(
		func(w http.ResponseWriter, r *http.Request) {
			w.WriteHeader(http.StatusNotFound)
		},
		func(w http.ResponseWriter, r *http.Request) {
			require.Fail(t, "unexpected call to endpoint")
		},
	)

	ctx, _ := newContext(ts)

	err := appStop(ctx, "/test", false)
	assert.EqualError(t, err, "app '/test' does not exist")
}

func TestAppStopForce(t *testing.T) {
	app := app{
		App: goMarathon.Application{
			ID:        testID,
			Instances: intPointer(2),
		},
	}
	deployment := goMarathon.DeploymentID{
		DeploymentID: "5ed4c0c5-9ff8-4a6f-a0cd-f57f59a34b43",
		Version:      "2015-09-29T15:59:51.164Z",
	}

	ts := router(
		func(w http.ResponseWriter, r *http.Request) { encode(w, app) },
		func(w http.ResponseWriter, r *http.Request) {
			assert.Contains(t, r.URL.Query(), "force")
			encode(w, deployment)
		},
	)

	ctx, _ := newContext(ts)

	err := appStop(ctx, "/test", true)
	assert.NoError(t, err)
}

func TestAppStopWhenDeploymentBlocked(t *testing.T) {
	app := app{
		App: goMarathon.Application{
			ID:        testID,
			Instances: intPointer(2),
		},
	}

	ts := router(
		func(w http.ResponseWriter, r *http.Request) { encode(w, app) },
		func(w http.ResponseWriter, r *http.Request) { w.WriteHeader(409) },
	)

	ctx, _ := newContext(ts)

	err := appStop(ctx, "/test", false)
	assert.EqualError(t, err, "changes blocked: deployment already in progress for app")
}

func encode(w http.ResponseWriter, obj interface{}) {
	json.NewEncoder(w).Encode(obj)
}

func router(getFunc http.HandlerFunc, putFunc http.HandlerFunc) *httptest.Server {
	return httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		switch url := r.URL.Path; url {
		case "/service/marathon/v2/apps/test":
			switch r.Method {
			case http.MethodGet:
				getFunc(w, r)
			case http.MethodPut:
				putFunc(w, r)
			default:
				panic("unexpected http verb " + r.Method)
			}
		default:
			panic("unexpected call to endpoint " + url)
		}
	}))
}

func newContext(ts *httptest.Server) (*mock.Context, *bytes.Buffer) {
	out := new(bytes.Buffer)
	env := mock.NewEnvironment()
	env.Out = out
	ctx := mock.NewContext(env)
	cluster := config.NewCluster(nil)
	cluster.SetURL(ts.URL)
	ctx.SetCluster(cluster)
	return ctx, out
}

```

Comment (janisz) [reply pkg/cmd/marathon/app/app_stop_test.go:134]: Expecially for @drewvanstone 

```go
package app

import (
	"bytes"
	"encoding/json"
	"errors"
	"net/http"
	"net/http/httptest"
	"testing"

	"github.com/dcos/dcos-cli/pkg/config"
	"github.com/dcos/dcos-cli/pkg/mock"
	goMarathon "github.com/gambol99/go-marathon"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

const testID = "test"

func TestAppStop(t *testing.T) {
	simpleApp := struct {
		App goMarathon.Application `json:"app"`
	}{App: goMarathon.Application{
		ID:        testID,
		Instances: intPointer(2),
	}}
	tests := []struct {
		name    string
		force   bool
		getFunc http.HandlerFunc
		putFunc http.HandlerFunc
		out     string
		err     error
	}{
		{
			getFunc: encodeFunc(simpleApp),
			putFunc: encodeFunc(goMarathon.DeploymentID{
				DeploymentID: "5ed4c0c5-9ff8-4a6f-a0cd-f57f59a34b43",
				Version:      "2015-09-29T15:59:51.164Z",
			}),
			out: "Created deployment 5ed4c0c5-9ff8-4a6f-a0cd-f57f59a34b43\n",
		},
		{
			name: "when app is stopped",
			getFunc: encodeFunc(struct {
				App goMarathon.Application `json:"app"`
			}{App: goMarathon.Application{
				ID:        testID,
				Instances: intPointer(0),
			}}),
			err: errors.New("app '/test' already stopped: 0 instances"),
		},
		{
			name:    "when app doesn't exist",
			getFunc: http.NotFound,
			err:     errors.New("app '/test' does not exist"),
		},
		{
			name:    "with force",
			force:   true,
			getFunc: encodeFunc(simpleApp),
			putFunc: func(w http.ResponseWriter, r *http.Request) {
				assert.Contains(t, r.URL.Query(), "force")
				encode(w, goMarathon.DeploymentID{
					DeploymentID: "5ed4c0c5-9ff8-4a6f-a0cd-f57f59a34b43",
					Version:      "2015-09-29T15:59:51.164Z",
				})
			},
			out: "Created deployment 5ed4c0c5-9ff8-4a6f-a0cd-f57f59a34b43\n",
		},
		{
			name:    "when deployment is blocked",
			getFunc: encodeFunc(simpleApp),
			putFunc: func(w http.ResponseWriter, r *http.Request) { w.WriteHeader(409) },
			err:     errors.New("changes blocked: deployment already in progress for app"),
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			ts := router(tt.getFunc, tt.putFunc)
			ctx, out := newContext(ts)

			err := appStop(ctx, "/test", tt.force)

			require.Equal(t, tt.err, err)
			assert.Equal(t, tt.out, out.String())
		})
	}
}

func encodeFunc(obj interface{}) func(w http.ResponseWriter, r *http.Request) {
	return func(w http.ResponseWriter, r *http.Request) { encode(w, obj) }
}

func encode(w http.ResponseWriter, obj interface{}) {
	json.NewEncoder(w).Encode(obj)
}

func router(getFunc http.HandlerFunc, putFunc http.HandlerFunc) *httptest.Server {
	return httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		switch url := r.URL.Path; url {
		case "/service/marathon/v2/apps/test":
			switch r.Method {
			case http.MethodGet:
				getFunc(w, r)
			case http.MethodPut:
				putFunc(w, r)
			default:
				panic("unexpected http verb " + r.Method)
			}
		default:
			panic("unexpected call to endpoint " + url)
		}
	}))
}

func newContext(ts *httptest.Server) (*mock.Context, *bytes.Buffer) {
	out := new(bytes.Buffer)
	env := mock.NewEnvironment()
	env.Out = out
	ctx := mock.NewContext(env)
	cluster := config.NewCluster(nil)
	cluster.SetURL(ts.URL)
	ctx.SetCluster(cluster)
	return ctx, out
}

```

**Review by janisz — APPROVED** [review]
LGTM

**Review by armandgrillet — APPROVED** [review]

File: pkg/cmd/marathon/app/app_stop_test.go [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+		}
+	}))
+
+	out := new(bytes.Buffer)
+	env := mock.NewEnvironment()
+	env.Out = out
+	ctx := mock.NewContext(env)
+	cluster := config.NewCluster(nil)
+	cluster.SetURL(ts.URL)
+	ctx.SetCluster(cluster)
```
Comment (janisz) [inline pkg/cmd/marathon/app/app_stop_test.go]: https://github.com/dcos/dcos-core-cli/pull/417#discussion_r350699733

**Review by janisz — CHANGES_REQUESTED** [review]
Take a look at https://github.com/dcos/dcos-core-cli/pull/417#discussion_r350699733

**Pull request merged**